// @splunk/otel/lib/instrumentations/external/neo4j/semconv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ATTR_DB_SYSTEM = exports.ATTR_DB_STATEMENT = exports.ATTR_DB_OPERATION = exports.ATTR_NET_TRANSPORT = exports.ATTR_NET_PEER_PORT = exports.ATTR_NET_PEER_NAME = exports.ATTR_DB_USER = exports.ATTR_DB_NAME = void 0;
/*
 * Copyright Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
exports.ATTR_DB_NAME = 'db.name';
exports.ATTR_DB_USER = 'db.user';
exports.ATTR_NET_PEER_NAME = 'net.peer.name';
exports.ATTR_NET_PEER_PORT = 'net.peer.port';
exports.ATTR_NET_TRANSPORT = 'net.transport';
exports.ATTR_DB_OPERATION = 'db.operation';
exports.ATTR_DB_STATEMENT = 'db.statement';
exports.ATTR_DB_SYSTEM = 'db.system';
